### PR TITLE
Monaco show basic suggestions for Twig env

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Monaco's support for Twig doesn't include any information about the default environment, possibly because it can change (although it doesn't explain why keywords are missing). This PR at least adds some of the more common, available by default, keywords, functions, filters, etc to the list.
Obviously, this is still no replacement for user's reading the Twig documentation, but may give some hints to them that they do indeed have access to a default Twig environment and have some idea of what is available to them.

### CODE REVIEW

**Review by cedric-anne — CHANGES_REQUESTED** [review]
User defined Twig templates should rely on the `Glpi\ContentTemplates\TemplateManager` security policy. Indeed, the rendering environment we use for our application templates is not safe at all, as, for instance, it exposes the `call` function that permit to call any PHP function, including those that can be used to read/write/exec any file accessible on the system.

If we have some user generated templates that are rendered using the `Glpi\Application\View\TemplateRenderer`, it has to be changed.

See the security policy:
https://github.com/glpi-project/glpi/blob/375e81f8ccecb5a16cb0f35b3cf24429fa701546/src/ContentTemplates/TemplateManager.php#L162-L170

**Comment by cconard96** [comment]
Nothing I added uses TemplateRenderer, but uses a new, default Twig environment with an array loader that has only the single user-defined template instead. I knew that user's shouldn't have any access to the added extensions we have. The default environment doesn't seem to allow any kind of arbitrary code execution unless we pass objects into the template and then they could access a limited set of data on the items.

The only difference I see between the current implementation and Glpi\ContentTemplates\TemplateManager is it isn't quite as restrictive with the filters/functions they can use (reverse, various array filters, some formatters, encoding filters, etc are not available through TemplateManager) and it doesn't get sanitized by `RichText::getSafeHtml` (not sure this is desired in some cases like webhooks).

**Comment by cconard96** [comment]
Should TemplateManager be made less restrictive and have the sanitization part be optional?

**Comment by cedric-anne** [comment]
> Nothing I added uses TemplateRenderer, but uses a new, default Twig environment with an array loader that has only the single user-defined template instead. I knew that user's shouldn't have any access to the added extensions we have. The default environment doesn't seem to allow any kind of arbitrary code execution unless we pass objects into the template and then they could access a limited set of data on the items.

OK, so it is less a problem.

> Should TemplateManager be made less restrictive

Yes. It can provide more features, as long as it does not provide, for instance, the `include` and the `extends` features that may permit unexpected access to some files.

> and have the sanitization part be optional?

It has been made optional recently.

File: js/modules/Monaco/MonacoEditor.js [hunk 1]
```diff
@@ -166,6 +169,62 @@ export default class MonacoEditor {
             });
         }
     }
+
+    getTwigCompletions() {
+        const keywords = [
+            // (opening) tags
+            'apply', 'autoescape', 'block', 'for', 'if', 'macro', 'set',
```
Comment (cedric-anne) [inline js/modules/Monaco/MonacoEditor.js:176]:
```suggestion
            'apply', 'autoescape', 'block', 'for', 'if', 'elseif', 'macro', 'set',
```

File: js/modules/Monaco/MonacoEditor.js [hunk 2]
```diff
@@ -166,6 +169,62 @@ export default class MonacoEditor {
...
+            // closing tags
+            'endapply', 'endautoescape', 'endblock', 'endfor', 'endif', 'endmacro', 'endset',
+            // literals
+            'true', 'false', 'null'
+        ];
+        const operators = [
+            'in', 'is', 'and', 'or', 'not', 'b-and', 'b-xor', 'b-or', 'matches', 'starts with', 'ends with', 'has some', 'has every',
+        ];
+        const functions = ['min', 'max', 'random', 'range'];
```
Comment (cedric-anne) [inline js/modules/Monaco/MonacoEditor.js:185]: It is available both in filters and functions.
```suggestion
        const functions = ['date', 'min', 'max', 'random', 'range'];
```

File: js/modules/Monaco/MonacoEditor.js [hunk 3]
```diff
@@ -166,6 +169,62 @@ export default class MonacoEditor {
...
+        // Like functions but occur after '|'
+        const filters = [
+            'abs', 'batch', 'capitalize', 'column', 'default', 'escape', 'filter', 'first', 'format', 'join',
```
Comment (cedric-anne) [inline js/modules/Monaco/MonacoEditor.js:188]:
```suggestion
            'abs', 'batch', 'capitalize', 'column', 'date', 'default', 'escape', 'filter', 'first', 'format', 'join',
```

**Review by cedric-anne — COMMENTED** [review]
Seems OK.

**Review by cedric-anne — APPROVED** [review]

**Review by AdrienClairembault — APPROVED** [review]

**Pull request merged**